### PR TITLE
[Serialization] Accept a raw pointer in serialization

### PR DESCRIPTION
**Pull request by Tishj**
This PR makes it so that in the serialization part of the code we accept a raw pointer (of T) if the type to serialize is a pointer.

I've only enabled this feature if `serialize_property` is provided.
The reasoning behind this is that we can expect the type to be the same between serialization and deserialization if `serialize_property` is not provided.

But if it's provided we are instead serializing the result of an expression.

The place I ran into this is with "optionally owned" pointers, such as:
```c++
	//! Owned column data collection, if any
	unique_ptr<ColumnDataCollection> collection;
	//! The column data collection to scan
	ColumnDataCollection &to_scan;
```

For which the serialization is:
```json
      {
        "id": 202,
        "name": "collection",
        "type": "ColumnDataCollection*",
        "serialize_property": "&to_scan"
      }
```

In this case, we **always** have a `to_scan`, before we serialize this is a reference to an existing ColumnDataCollection (which we optionally own).
When we need to serialize and deserialize, we can no longer refer to this existing ColumnDataCollection if we don't own it, so we serialize the ColumnDataCollection to disk (whether we own it or not!)

And when we deserialize we deserialize it into `collection`, so after deserialization we always own our column data collection.

### CODE REVIEW

File: src/storage/serialization/serialize_nodes.cpp [hunk 1]
```diff
@@ -309,7 +309,7 @@ JoinCondition JoinCondition::Deserialize(Deserializer &deserializer) {
 
 void LogicalType::Serialize(Serializer &serializer) const {
 	serializer.WriteProperty<LogicalTypeId>(100, "id", id_);
-	serializer.WritePropertyWithDefault<shared_ptr<ExtraTypeInfo>>(101, "type_info", type_info_);
+	serializer.WritePropertyWithDefault<ExtraTypeInfo*>(101, "type_info", type_info_.get());
```
Comment (Mytherin) [inline src/storage/serialization/serialize_nodes.cpp:312]: Are these changes necessary?

File: src/include/duckdb/common/serializer/serialization_traits.hpp [hunk 2]
```diff
@@ -124,6 +124,12 @@ struct is_optional_ptr<optional_ptr<T>> : std::true_type {
 	typedef T ELEMENT_TYPE;
 };
 
+template <typename T>
+struct is_pointer : std::false_type {};
+
+template <typename T>
+struct is_pointer<T *> : std::true_type {};
```
Comment (Mytherin) [inline src/include/duckdb/common/serializer/serialization_traits.hpp:131]: Can this not be either an `optional_ptr`, or a reference?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Can we try to add support for serializing a reference instead perhaps?

File: src/include/duckdb/common/serializer/serialization_traits.hpp (hunk 2 above)
Comment (Tishj) [reply src/include/duckdb/common/serializer/serialization_traits.hpp:131]: I can turn it into a reference 👍 
If I turn it into an `optional_ptr` I am just shifting the problem, because I can already create an unowned unique_ptr on the spot (which is medium hacky), but it takes a `const &`, not a `&&` so that didn't work

File: src/storage/serialization/serialize_nodes.cpp (hunk 1 above)
Comment (Tishj) [reply src/storage/serialization/serialize_nodes.cpp:312]: The change to `ExtraTypeInfo*` from `shared_ptr<ExtraTypeInfo>` yes, if I have an optional_ptr or used the referencing operator on a reference (`&to_scan` for example) it can't translate to a `shared_ptr`.

This class got caught in the crossfire because it met the conditions: `is_pointer` and `custom_serialize_property`

File: src/include/duckdb/common/serializer/serialization_traits.hpp (hunk 2 above)
Comment (Tishj) [reply src/include/duckdb/common/serializer/serialization_traits.hpp:131]: Though if we accept a reference, then we can't serialize a null anymore, which is not the intention here

**Comment by Tishj** [comment]
```
Query unexpectedly failed (test/extension/test_alias_point.test:30)
================================================================================
SELECT add_point(({'x': 2, 'y': 3})::POINT, ({'x': 3, 'y': 4})::POINT);
================================================================================
Actual result:
Serialization Error: Failed to deserialize: field id mismatch, expected: 100, got: 65535
```

I have some fixing to do it seems

File: src/storage/serialization/serialize_logical_operator.cpp [hunk 3]
```diff
@@ -335,7 +335,7 @@ unique_ptr<LogicalOperator> LogicalCreateIndex::Deserialize(Deserializer &deseri
 
 void LogicalCreateTable::Serialize(Serializer &serializer) const {
 	LogicalOperator::Serialize(serializer);
-	serializer.WritePropertyWithDefault<unique_ptr<CreateInfo>>(200, "info", info->base);
+	serializer.WritePropertyWithDefault<optional_ptr<CreateInfo>>(200, "info", info->base);
```
Comment (Mytherin) [inline src/storage/serialization/serialize_logical_operator.cpp:338]: Can we avoid changing these?

File: scripts/generate_serialization.py [hunk 4]
```diff
@@ -586,7 +597,12 @@ def generate_class_code(class_entry):
             deserialize_template_str = deserialize_element_class_base.replace(
                 '${BASE_PROPERTY}', entry.base.replace('*', '')
             ).replace('${DERIVED_PROPERTY}', entry.type.replace('*', ''))
-        type_name = replace_pointer(entry.type)
+        if entry.custom_serialize_property == True and is_pointer(entry.type):
```
Comment (Mytherin) [inline scripts/generate_serialization.py:600]: I don't quite follow the need for this extra code here - why turn entry types that are not marked as an optional ptr into an optional ptr here?

Comment (Tishj) [reply scripts/generate_serialization.py:600]: Perhaps it's necessary to provide further context:
<https://github.com/Tishj/duckdb/pull/129>
This is the PR that I need this for, I figured this was stand alone and perhaps controversial (my hunch was right) enough to be a separate PR

Comment (Tishj) [reply scripts/generate_serialization.py:600]: So I talked about it a bit with Max, Carlo and Sam and they proposed a good idea: `optionally_owned_ptr`
That special cases this relationship in a way that we probably don't have to touch other serialization code to support it

Basically it encapsulates these two variables we have here:
`owned`
`ref`

If it's constructed with a unique_ptr, it sets owned and ref, if it's constructed with a reference, it only sets ref

**Comment by Tishj** [comment]
I think we can close this, I will be adding an `optionally_owned_ptr` in a future PR instead to solve this
